File: haskell/day8.hs
import qualified Data.Set as Set
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

parseLine :: T.Text -> (String, Int)
parseLine line = (ins, param)
    where
        (insT:paramT:_) = T.words line
        ins = T.unpack insT
        param = read $ T.unpack $ T.replace (T.pack "+") T.empty paramT

parse :: T.Text -> [(String, Int)]
parse content = map parseLine $ T.lines $ content

doInstr :: (String, Int) -> Int -> Int -> (Int, Int)

doInstr ("acc", param) pc acc = (pc + 1, acc + param)

doInstr ("jmp", param) pc acc = (pc + param, acc)

doInstr ("nop", param) pc acc = (pc + 1, acc)

execute :: [(String, Int)] -> Set.Set Int -> Int -> Int -> Int
execute instructions seenPCs pc acc =
    if Set.member pc seenPCs then
        acc
    else
        execute instructions newSeenPCs newPC newAcc
    where
        newSeenPCs = Set.insert pc seenPCs
        (newPC, newAcc) = doInstr (instructions !! pc) pc acc

answer :: [(String, Int)] -> Int
answer content =
    execute content Set.empty 0 0

main :: IO ()
main = do
    content <- TIO.readFile "../input/day8.txt"
    let input = parse content
    putStrLn $ show $ answer input

File: c++/day8.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>

struct State {
    int pc;
    int acc;
};

class Instruction {
public:
    Instruction(int param) : m_param(param) {}
    virtual ~Instruction() {}
    virtual void execute(State &state) = 0;
protected:
    int m_param;
};

class Accumulate : public Instruction {
    using Instruction::Instruction;

    void execute(State &state) override {
        state.acc += m_param;
        state.pc += 1;
    }
};

class Jump : public Instruction {
    using Instruction::Instruction;

    void execute(State &state) override {
        state.pc += m_param;
    }
};


class Nop : public Instruction {
    using Instruction::Instruction;

    void execute(State &state) override {
        state.pc += 1;
    }
};

using Program = std::vector<Instruction *>;

Program readParseInput(std::string fileName) {
    std::ifstream file(fileName);
    Program program;

    std::string line;
    while (std::getline(file, line)) {
        int param = std::atoi(&line[4]);
        Instruction *instruction = 0;
        if (line.starts_with("acc")) {
            instruction = new Accumulate(param);
        } else if (line.starts_with("jmp")) {
            instruction = new Jump(param);
        } else {
            instruction = new Nop(param);
        }
        program.push_back(instruction);
    }
    file.close();

    return program;
}

int answer(const Program &program) {
    State s = {0, 0};
    std::set<int> seenPcs;

    while (true) {
        if (seenPcs.find(s.pc) != seenPcs.end())
            return s.acc;
        seenPcs.insert(s.pc);
        program[s.pc]->execute(s);
    }
}

int main(void) {
    std::cout << answer(readParseInput("../input/day8.txt")) << '\n';
    return 0;
}
